### ISSUE TITLE
[Monitoring] Prometheus Endpoint

### ISSUE THREAD

**Issue by DanielFuchs98** [dswd/vpncloud#57]
It would be nice if the VpnCloud Stats could be exported via a Prometheus-Endpoint.

**Comment by dswd** [comment]
As far as I understand such an endpoint should be a small HTTP server. Adding such a server to vpncloud would add a lot of dependencies and some code that only exists to provide this endpoint. I could feature-gate this functionality but then you would need to compile vpncloud for yourself. Alternatively, I could add an option to export the stats in json or maybe yaml. Then you would need to setup a webserver to provide the file to prometheus.
What are thoughts on that?

**Comment by DanielFuchs98** [comment]
Sounds better - I think only a few users would use the prometheus endpoint, it isn't worth it.

VPNCloud has already an stats file, maybe you can implement a setting for the output format of this file. In my opinion there are three possible options: Text (actual stats-file), JSON and YAML.

What do you think?

**Comment by night** [comment]
alternatively, maybe worth throwing in statsd support instead, which just emits to a udp server (usually on localhost) if running a preconfigured set of metrics. this might be preferred since most monitoring solutions seem to have an ingests for this.

**Comment by dswd** [comment]
Actually that sounds reasonable. How about proceeding with the following:

I can change the format of the stats file to be YAML. This would both be parsable and still human-readable. I think supporting multiple stats output formats would be too much functionality for this use case.

The statsd format is so simple, this can be implemented fairly easily. So I would also implement this.

Let's do some brainstorming: What kind of values are important for monitoring?

Here is what I came up with:
- Number of peers
- Inbound traffic (bytes and packets)
- Outbound traffic (bytes and packets)

**Comment by DanielFuchs98** [comment]
Yes, your metrics are good. Important are also package drops/errors.

**Comment by dswd** [comment]
@DanielFuchs98 @night please check the latest commit.
I changed the stats format to yaml that is still readable.
Also I added statsd support. Please check that in particular, as I do not have any statsd server.

**Comment by DanielFuchs98** [comment]
Thanks for your quick work! I´ll give it a try with the Prometheus-Statsd-Exporter (https://github.com/prometheus/statsd_exporter), it look promising.

**Comment by night** [comment]
the statsd implementation may need a bit more effort to make it more consumable.. i actually might recommend using a library for it (even though the protocol is simple). it looks like you may have gotten the protocol wrong too (https://github.com/dswd/vpncloud/commit/1f94d0deffcc8b08b4eb1789ea2936e49bb2a057#diff-12acc60ff6b86016b459365061e40533R605 and below seem to be missing a metric type)

maybe something like https://github.com/tshlabs/cadence would be an interesting lib, which also supports tagging. Tagging would allow you to, for example, tag metrics by peer or classify the metrics by the network (if you have multiple networks running concurrently). if no tagging is added, the ability to prefix metric names would be a great alternative.

**Comment by dswd** [comment]
I updated the code. I also added a prefix option.
If you think its worth it, I can add per-peer and per-connection data as well. However that would spam the statsd buckets as those keys will change over time.

**Comment by DanielFuchs98** [comment]
In my lab-env i´ve setup vpncloud on one host and tried to ping an not existing ip (same subnet). As expected i got a "no route", but the metrics dont show this.

# Commands:
```
root@l0001:/home/fuchsd013# ./vpncloud --statsd-server localhost:9125 --ifup 'ifconfig $IFNAME 10.0.0.1/24 mtu 1400 up' &
[1] 7725
root@l0001:/home/fuchsd013# ip a
...
59: vpncloud0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc pfifo_fast state UNKNOWN group default qlen 1000
    link/ether 32:74:20:b7:81:d1 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.1/24 brd 10.0.0.255 scope global vpncloud0
       valid_lft forever preferred_lft forever

root@l0001:/home/fuchsd013# curl 10.0.0.5
curl: (7) Failed to connect to 10.0.0.5 port 80: Keine Route zum Zielrechner
root@l0001:/home/fuchsd013# curl 10.0.0.5
curl: (7) Failed to connect to 10.0.0.5 port 80: Keine Route zum Zielrechner
```

# Metrics:
```
vpncloud_dropped_payload_bytes 0
vpncloud_dropped_payload_packets 0
vpncloud_invalid_protocol_traffic_bytes 0
vpncloud_invalid_protocol_traffic_packets 0
vpncloud_peer_count 0
vpncloud_table_entries 0
vpncloud_traffic_payload_inbound_bytes 0
vpncloud_traffic_payload_inbound_packets 0
vpncloud_traffic_payload_outbound_bytes 1134
vpncloud_traffic_payload_outbound_packets 27
vpncloud_traffic_protocol_inbound_bytes 0
vpncloud_traffic_protocol_inbound_packets 0
vpncloud_traffic_protocol_outbound_bytes 0
vpncloud_traffic_protocol_outbound_packets 0
```

I has expected, that vpncloud_dropped_payload_bytes are greater than zero. Except that, its looks very good to me.

**Comment by DanielFuchs98** [comment]
Can you add tags to the metrics? For example the VPN-IP, so it would be simpler to separate that. There are three possible formats: https://github.com/prometheus/statsd_exporter/blob/master/README.md#tagging-extensions

**Comment by dswd** [comment]
@DanielFuchs98 
1) The "no route" not showing up in vpncloud_dropped_payload_bytes looks unintuitive at first but is  correct behavior when using TAP devices (i.e. ethernet). When you ping an address, an ARP request for that unknown IP is broadcast (via vpncloud) and for an unknown IP, no reply is received (but nothing is ever dropped). After a short timeout, the kernel decides to identify this IP as "no route". If you use a TUN device, you should be able to see dropped packets as this time it is vpncloud that decides that the host is unknown.

2) I added a prefix option so you can define the statsd prefix per instance (which defaults to "vpncloud"). I considered using the hostname as default but lots of nodes have a meaningless hostname set, so "vpncloud" is actually better. If you need some more options, just propose something.

**Comment by DanielFuchs98** [comment]
Works, thanks for your quick help! Are you planning to push the update in the Debian-Repo?

**Comment by dswd** [comment]
I just released 1.4.0 and will push that to the repo in a few days
